## Classification ในการจำแนกคนที่เป็นโรคหัวใจกับไม่เปนโรคหัวใจ โดยมีตัวแปร 13 Column

In [3]:
# import library เพื่อทำ data manipulation
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
# อ่านข้อมูล CSV เป้น Dataframe
df = pd.read_csv('Classification.csv')

In [5]:
# เปลี่ยนชื่อ column เพื่อง่ายแก่การตีความ
df.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

In [6]:
df.sample(10)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
32    44    1   1       130   219    0        0      188      0      0.0   
137   62    1   1       128   208    1        0      140      0      0.0   
258   62    0   0       150   244    0        1      154      1      1.4   
204   62    0   0       160   164    0        0      145      0      6.2   
200   44    1   0       110   197    0        0      177      0      0.0   
150   66    1   0       160   228    0        0      138      0      2.3   
143   67    0   0       106   223    0        1      142      0      0.3   
75    55    0   1       135   250    0        0      161      0      1.4   
81    45    1   1       128   308    0        0      170      0      0.0   
58    34    1   3       118   182    0        0      174      0      0.0   

     slope  ca  thal  target  
32       2   0     2       1  
137      2   0     2       1  
258      1   0     2       0  
204      0   3     3       0  
200      2   1     2       0  
150      2   0     1       1  
143      2   2     2       1  
75       1   0     2       1  
81       2   0     2       1  
58       2   0     2       1

In [7]:
df.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [8]:
# ดูว่าข้อมูลมีค่า Null ในแต่ละแถวไหม
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [9]:
df.dtypes

age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca            int64
thal          int64
target        int64
dtype: object

In [10]:
X = df.iloc[:, :-1].values
y = df.iloc[:, -1].values

In [11]:
X.shape , y.shape

((303, 13), (303,))

In [12]:
X, y

(array([[63.,  1.,  3., ...,  0.,  0.,  1.],
        [37.,  1.,  2., ...,  0.,  0.,  2.],
        [41.,  0.,  1., ...,  2.,  0.,  2.],
        ...,
        [68.,  1.,  0., ...,  1.,  2.,  3.],
        [57.,  1.,  0., ...,  1.,  1.,  3.],
        [57.,  0.,  1., ...,  1.,  1.,  2.]]),
 array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

# Train/Test Split

In [13]:
from sklearn.model_selection import train_test_split
X_train , X_test , y_train , y_test = train_test_split(X,y, test_size=0.2, random_state=0)

In [14]:
X_train , X_train.shape

(array([[43.,  0.,  2., ...,  1.,  0.,  2.],
        [66.,  0.,  2., ...,  1.,  1.,  2.],
        [58.,  1.,  2., ...,  2.,  0.,  2.],
        ...,
        [56.,  1.,  3., ...,  1.,  0.,  3.],
        [47.,  1.,  2., ...,  2.,  0.,  2.],
        [58.,  1.,  1., ...,  1.,  0.,  2.]]),
 (242, 13))

# Decision Tree Classifier without normalize

In [15]:
# import library Decision tree with criterion entropy
clf = DecisionTreeClassifier(criterion = 'entropy',random_state=0)
clf.fit(X_train,y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [16]:
print(f'Accuracy on training set: {clf.score(X_train,y_train)} , Accuracy on test set : {clf.score(X_test,y_test)}')

Accuracy on training set: 1.0 , Accuracy on test set : 0.8032786885245902


# Decision Tree Classifier with Std. Normalize

In [17]:
X = df.iloc[:, :-1].values
y = df.iloc[:, -1].values

In [18]:
from sklearn.model_selection import train_test_split
X_train , X_test , y_train , y_test = train_test_split(X,y, test_size=0.2, random_state=0)

### Feature scaling only value that not categorical features

In [19]:
# Features scaling using std
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train[:,[0,3,4,7,9]] = sc_X.fit_transform(X_train[:,[0,3,4,7,9]])
X_test[:,[0,3,4,7,9]] = sc_X.fit_transform(X_test[:,[0,3,4,7,9]])

In [20]:
# check data that norm completely
X_train[:,[0,3,4,7,9]]

array([[-1.32773282, -0.57412513, -0.63267424,  0.65626162, -0.72460883],
       [ 1.24903178,  0.83106608,  0.58543663,  0.09400698, -0.89249331],
       [ 0.35276583,  0.47976828, -0.67015457,  0.65626162, -0.89249331],
       ...,
       [ 0.12869935, -0.69122439, -1.00747758,  0.52651055,  0.70240923],
       [-0.87959984,  0.36266901,  0.19189312,  0.2670084 , -0.89249331],
       [ 0.35276583, -0.69122439,  0.69787763,  0.44000983,  0.61846699]])

In [21]:
X_train, X_train.shape

(array([[-1.32773282,  0.        ,  2.        , ...,  1.        ,
          0.        ,  2.        ],
        [ 1.24903178,  0.        ,  2.        , ...,  1.        ,
          1.        ,  2.        ],
        [ 0.35276583,  1.        ,  2.        , ...,  2.        ,
          0.        ,  2.        ],
        ...,
        [ 0.12869935,  1.        ,  3.        , ...,  1.        ,
          0.        ,  3.        ],
        [-0.87959984,  1.        ,  2.        , ...,  2.        ,
          0.        ,  2.        ],
        [ 0.35276583,  1.        ,  1.        , ...,  1.        ,
          0.        ,  2.        ]]),
 (242, 13))

In [22]:
from sklearn.tree import DecisionTreeClassifier
clf2 = DecisionTreeClassifier(criterion = 'entropy',random_state=0)
clf2.fit(X_train,y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [23]:
print(f'Accuracy on training set: {clf2.score(X_train,y_train)} , Accuracy on test set : {clf2.score(X_test,y_test)}')

Accuracy on training set: 1.0 , Accuracy on test set : 0.8032786885245902


In [24]:
from sklearn.tree import DecisionTreeClassifier
clf3 = DecisionTreeClassifier(criterion = 'gini',random_state=0)
clf3.fit(X_train,y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [25]:
print(f'Accuracy on training set: {clf3.score(X_train,y_train)} , Accuracy on test set : {clf3.score(X_test,y_test)}')

Accuracy on training set: 1.0 , Accuracy on test set : 0.7704918032786885


### Decision Tree Classifier with Min-Max Scaler

In [26]:
X = df.iloc[:, :-1].values
y = df.iloc[:, -1].values

In [27]:
from sklearn.model_selection import train_test_split
X_train , X_test , y_train , y_test = train_test_split(X,y, test_size=0.2, random_state=0)

In [28]:
# Features scaling using min-max
from sklearn.preprocessing import MinMaxScaler
min_max_scaler = MinMaxScaler()
X_train[:,[0,3,4,7,9]] = min_max_scaler.fit_transform(X_train[:,[0,3,4,7,9]])
X_test[:,[0,3,4,7,9]] = min_max_scaler.fit_transform(X_test[:,[0,3,4,7,9]])

In [29]:
X_train[:,[0,3,4,7,9]]

array([[0.29166667, 0.28571429, 0.18937644, 0.71755725, 0.03225806],
       [0.77083333, 0.53061224, 0.33949192, 0.61832061, 0.        ],
       [0.60416667, 0.46938776, 0.18475751, 0.71755725, 0.        ],
       ...,
       [0.5625    , 0.26530612, 0.14318707, 0.69465649, 0.30645161],
       [0.375     , 0.44897959, 0.29099307, 0.64885496, 0.        ],
       [0.60416667, 0.26530612, 0.35334873, 0.67938931, 0.29032258]])

In [30]:
X_train ,X_train.shape

(array([[0.29166667, 0.        , 2.        , ..., 1.        , 0.        ,
         2.        ],
        [0.77083333, 0.        , 2.        , ..., 1.        , 1.        ,
         2.        ],
        [0.60416667, 1.        , 2.        , ..., 2.        , 0.        ,
         2.        ],
        ...,
        [0.5625    , 1.        , 3.        , ..., 1.        , 0.        ,
         3.        ],
        [0.375     , 1.        , 2.        , ..., 2.        , 0.        ,
         2.        ],
        [0.60416667, 1.        , 1.        , ..., 1.        , 0.        ,
         2.        ]]),
 (242, 13))

In [31]:
from sklearn.tree import DecisionTreeClassifier
clf4 = DecisionTreeClassifier(criterion = 'entropy',random_state=0)
clf4.fit(X_train,y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [32]:
print(f'Accuracy on training set: {clf4.score(X_train,y_train)} , Accuracy on test set : {clf4.score(X_test,y_test)}')

Accuracy on training set: 1.0 , Accuracy on test set : 0.7540983606557377


In [33]:
from sklearn.tree import DecisionTreeClassifier
clf5 = DecisionTreeClassifier(criterion = 'gini',random_state=0)
clf5.fit(X_train,y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [34]:
print(f'Accuracy on training set: {clf5.score(X_train,y_train)} , Accuracy on test set : {clf5.score(X_test,y_test)}')

Accuracy on training set: 1.0 , Accuracy on test set : 0.7377049180327869


# Random forest without Normalize using Critetion = 'Gini'

In [35]:
X = df.iloc[:, :-1].values
y = df.iloc[:, -1].values

In [36]:
from sklearn.model_selection import train_test_split
X_train , X_test , y_train , y_test = train_test_split(X,y, test_size=0.2, random_state=0)

In [37]:
from sklearn.ensemble import RandomForestClassifier
n_list = [5,10,100,300,500,1000]
MaxVars = 0
number = 0
for n in n_list:
    clf6 = RandomForestClassifier(n_estimators=n, random_state=0, criterion='gini')
    clf6.fit(X_train,y_train)
    if clf6.score(X_test,y_test) > MaxVars:
        MaxVars = clf6.score(X_test,y_test)
        number = n
print(f'Number of Estimator : {number}')
print(f'Accuracy on test set : {MaxVars}')

Number of Estimator : 10
Accuracy on test set : 0.8852459016393442


## Random forest without Normalize using Critetion = 'Entropy'

In [38]:
X = df.iloc[:, :-1].values
y = df.iloc[:, -1].values

In [39]:
from sklearn.model_selection import train_test_split
X_train , X_test , y_train , y_test = train_test_split(X,y, test_size=0.2, random_state=0)

In [40]:
n_list = [5,10,100,300,500,1000]
MaxVars = 0
number = 0

for n in n_list:
    clf7 = RandomForestClassifier(n_estimators=n, random_state=0, criterion='entropy')
    clf7.fit(X_train,y_train)
    if clf7.score(X_test,y_test) > MaxVars:
        MaxVars = clf7.score(X_test,y_test)
        number = n
print(f'Number of Estimator : {number}')
print(f'Accuracy on test set : {MaxVars}')

Number of Estimator : 5
Accuracy on test set : 0.8524590163934426


In [41]:
clf7

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='entropy', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=1000,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

### Visualizing Tree

In [59]:
from sklearn.tree import export_graphviz

from https://towardsdatascience.com/how-to-visualize-a-decision-tree-from-a-random-forest-in-python-using-scikit-learn-38ad2d75f21c

In [74]:
model = RandomForestClassifier(max_depth=5)
model.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=5, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [84]:
estimator = model.estimators_[1]
feature_names = [i for i in df.columns[:-1]]

y_train_str = y_train.astype('str')
y_train_str[y_train_str == '0'] = 'no disease'
y_train_str[y_train_str == '1'] = 'disease'
y_train_str[:3]

array(['disease', 'disease', 'disease'], dtype='<U21')

In [86]:
from subprocess import call

In [90]:
export_graphviz(estimator, out_file='tree.dot', 
                feature_names = feature_names,
                class_names = y_train_str,
                rounded = True, proportion = False, 
                precision = 2, filled = True)

call(['dot', '-Tpng', 'C:/Users/theka/OneDrive/Desktop/code/Untitled Folder/project/tree.dot', '-o', '/tree.png', '-Gdpi=600'])

from IPython.display import Image
Image(filename = 'tree.png')

FileNotFoundError: [WinError 2] The system cannot find the file specified

### Compare Accuracy between Gini and Entropy

### เลือก Gini เพราะว่าค่า Accuracy ของ Gini ได้เท่ากับ 88.52% ซึ่งมากกว่า Entropy ที่ได้ 85.24%